From f4c5b04148db5fc4e5265beec461d3b7d9f4a789 Mon Sep 17 00:00:00 2001
From: Alexandre Julliard <julliard@winehq.org>
Date: Tue, 18 Aug 2026 12:45:10 +0200
Subject: [PATCH] ntdll: Reserve some space for top-down allocation in large
 address aware mode.

Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=58698
Upstream: https://gitlab.winehq.org/wine/wine/-/commit/f4c5b04148db5fc4e5265beec461d3b7d9f4a789
Backport adaptation: preserve CrossOver force_laa() by using its existing
large_address_space_active condition for the WoW64 branch.

---
 dlls/ntdll/unix/virtual.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/dlls/ntdll/unix/virtual.c b/dlls/ntdll/unix/virtual.c
index ca6327a59da4..5d1b58dc845a 100644
--- a/dlls/ntdll/unix/virtual.c
+++ b/dlls/ntdll/unix/virtual.c
@@ -5101,7 +5101,13 @@ void virtual_set_large_address_space(void)
                 free_reserved_memory( 0, (char *)0x7ffe0000 );
 #endif
         }
-        else user_space_wow_limit = (large_address_space_active ? limit_4g : limit_2g) - 1;
+        else if (large_address_space_active)
+        {
+            user_space_wow_limit = limit_4g - 1;
+            /* reserve space for top-down allocations; some apps break if the entire high 2G is available */
+            reserve_area( (void *)0xfff00000, (void *)0xffff0000 );
+        }
+        else user_space_wow_limit = limit_2g - 1;
     }
     else
     {
